### DISCUSSION TITLE
How to enable cache-level statistics

### DISCUSSION THREAD

**Discussion by vmgbrown** [infinispan/infinispan#10863] [Q&A] [answered]
Following up on my question from last week - I have been able to successfully enable and view cache-level statistics via the JMX console, but I still can't figure out how to do it via configuration. 

I was referred to the `statistics` attribute in this document: https://docs.jboss.org/infinispan/12.0/configdocs/infinispan-config-12.0.html

However, I don't know how to set that value in WildFly (v26), which only appears to provide a `statistics-enabled` attribute. Here is my current configuration, from `standalone.xml`:

```xml
<cache-container name="..." marshaller="JBOSS" modules="org.infinispan.hibernate-cache" statistics-enabled="true">
    <transport lock-timeout="60000"/>
    <local-cache name="local-query">
        <heap-memory size="10000"/>
        <expiration max-idle="100000"/>
    </local-cache>
    <invalidation-cache name="entity" remote-timeout="120000" statistics-enabled="true">
        <heap-memory size="100"/>
        <expiration max-idle="300000" lifespan="300000"/>
    </invalidation-cache>
    <replicated-cache name="timestamps" remote-timeout="120000">
        <expiration interval="0"/>
    </replicated-cache>
</cache-container>
```
Although I have `statistics-enabled` set to "true" for my "entity" cache, the value of `statisticsEnabled` is still `false` in JMX when I start my server.

**Comment by vmgbrown** [comment]
Does anyone have any thoughts on this? I am currently manually enabling the entity caches via JMX, but this is cumbersome since I am troubleshooting multiple caches and have to restart my server often.

**Comment by vmgbrown** [comment]
Another related question. In my cache container configuration, I have `heap-memory size="100"`. However, in my actual entity cache, I see via JMX that `evictionSize` is 10,000. I assumed it would be 100. Where is eviction size set?

**Reply by vmgbrown** [reply]
I just noticed that my "hibernate" cache container does set the heap memory size to 10,000:

```xml
<cache-container name="hibernate" marshaller="JBOSS" modules="org.infinispan.hibernate-cache">
    <transport lock-timeout="60000"/>
        <local-cache name="local-query">
            <heap-memory size="10000"/>
            <expiration max-idle="100000"/>
        </local-cache>
        <invalidation-cache name="entity" remote-timeout="120000">
            <heap-memory size="10000"/>
            <expiration max-idle="120000" lifespan="600000"/>
        </invalidation-cache>
        <replicated-cache name="timestamps" remote-timeout="120000">
            <expiration interval="0"/>
        </replicated-cache>
</cache-container>
```
Is this where the L2 cache gets the value?

Apologies for all the questions, but I have not found any documentation that explains this configuration.

**Reply by pferraro** [reply]
Yes, heap-memory size is corresponds to evictionSize.

**Reply by vmgbrown** [reply]
Thanks @pferraro - just to be clear, is the eviction size determined from the "hibernate" cache container or my named cache container? My named container declares a size of 100 (see above), but at runtime the cache has a size of 10,000 - which appears to have come from the "hibernate" container.

**Reply by pferraro** [reply]
Oh - are you trying to use a provide 2LC configuration from a different cache-container?  If so, you need to override the `<property name="hibernate.cache.infinispan.container">...</property>` in your persistence.xml.  Otherwise, this defaults to the "hibernate" cache-container.

**Reply by vmgbrown** [reply]
We actually have several cache containers, in addition to the default "hibernate" container. This configuration pre-dates my involvement in the project, so I'm not sure why it was done this way.

Is it not possible to use 2LC in multiple cache containers? Do the second-level caches just use the configuration in the "hibernate" container and ignore any configuration in the other containers?

**Reply by vmgbrown** [reply]
@pferraro Turns out our code is programmatically setting "hibernate.cache.infinispan.container" in some cases. Again, not sure why it was done this way, but it explains why the cache I'm investigating uses the default "hibernate" container while others are using different containers.

Thanks again for your help!

**Reply by pferraro** [reply]
> Is it not possible to use 2LC in multiple cache containers?

The 2LC can use separate cache-containers per persistent unit via the `hibernate.cache.infinispan.container` property.

**Comment by pferraro** [comment]
> Following up on my question from last week - I have been able to successfully enable and view cache-level statistics via the JMX console, but I still can't figure out how to do it via configuration.
> 
> I was referred to the `statistics` attribute in this document: https://docs.jboss.org/infinispan/12.0/configdocs/infinispan-config-12.0.html
> 
> However, I don't know how to set that value in WildFly (v26), which only appears to provide a `statistics-enabled` attribute. Here is my current configuration, from `standalone.xml`:
> 
> ```
> <cache-container name="..." marshaller="JBOSS" modules="org.infinispan.hibernate-cache" statistics-enabled="true">
>     <transport lock-timeout="60000"/>
>     <local-cache name="local-query">
>         <heap-memory size="10000"/>
>         <expiration max-idle="100000"/>
>     </local-cache>
>     <invalidation-cache name="entity" remote-timeout="120000" statistics-enabled="true">
>         <heap-memory size="100"/>
>         <expiration max-idle="300000" lifespan="300000"/>
>     </invalidation-cache>
>     <replicated-cache name="timestamps" remote-timeout="120000">
>         <expiration interval="0"/>
>     </replicated-cache>
> </cache-container>
> ```
> 
> Although I have `statistics-enabled` set to "true" for my "entity" cache, the value of `statisticsEnabled` is still `false` in JMX when I start my server.

Do you actually have a JPA/Hibernate application deployed with its second-level cache enabled?  You won't see any mbeans in JMX until the the second-level caches associated with the entity cache region have started.

**Reply by vmgbrown** [reply]
Yes, L2 caches are enabled. As I mentioned, I can manually enable statistics for these caches via JMX, and I can see the collected values at runtime. However, I can't figure out how to enable statistics via configuration.

**Reply by pferraro** [reply]
The configuration from your initial post looks fine to me.  I've verified it locally (using WF28).
Your Infinispan cache mbeans should be registered at: org.wildfly.clustering.infinispan:Cache=*

**Reply by pferraro** [reply]
Ah - I forgot about this: https://issues.redhat.com/browse/WFLY-15433
In WF26, you need to make sure to include this in your persistence.xml:
```xml
<property name="hibernate.cache.infinispan.statistics">true</property>
```
This was fixed in WF27.

**Reply by vmgbrown** [reply]
That's great, I'll add that now. Thanks!

**Reply by vmgbrown** [reply]
Just to confirm, do I need this property _and_ the `statistics-enabled` attribute in `standalone.xml`, or just this property?

**Reply by pferraro** [reply, marked as answer]
For WF26, you need both.  As of WF27, the hibernate.cache.infinispan.statistics property will be set automatically.

**Reply by vmgbrown** [reply]
Excellent, thanks for confirming.
